### PR TITLE
Preserve isolation level when retrying MySQL transactions

### PR DESCRIPTION
**Pull request by casperisfine**
Fix: https://github.com/rails/rails/pull/52501

Since MySQL require to set the isolation level as a statement before the `BEGIN`, if the `begin` statement fails and we retry it we're opening a transaction with the default level.

To handle this we can send both statement at once so that if we retry both are sent.

A much simpler alternative would be to just disallow retrying `BEGIN` when opening an isolated DB transaction.

cc @kirs 

Also cc @matthewd as I'd love your opinion on this. In the future I'd really like to leverage batches more to save on roundtrips as explained in https://github.com/trilogy-libraries/trilogy/issues/189, but right now it's actually worse because instead of 2 roundtrips we're doing 3:

  - `set_server_options OPTION_MULTI_STATEMENTS_ON`
  - `query "SET TRANSACTION ISOLATION LEVEL...; BEGIN;`
  - `set_server_options OPTION_MULTI_STATEMENTS_OFF`

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 1]
```diff
@@ -565,7 +569,7 @@ def table_options(table_name) # :nodoc:
 
       # SHOW VARIABLES LIKE 'name'
       def show_variable(name)
-        query_value("SELECT @@#{name}", "SCHEMA")
+        query_value("SELECT @@#{name}", "SCHEMA", materialize_transactions: false, allow_retry: true)
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:572]: I actually wonder if `query_value` shouldn't default to `materialize_transactions: false, allow_retry: true`, I'd need to review all of usages, but it's normally mean to read connection variables and schema, and using `query_value` to do a write would be weird...

**Comment by kirs** [comment]
Thank you! Something like this should certainly help us.

File: activerecord/test/cases/adapters/abstract_mysql_adapter/transaction_test.rb [hunk 2]
```diff
@@ -149,5 +149,32 @@ class Sample < ActiveRecord::Base
...
+        if sql.include?("BEGIN") && !first_begin_failed
+          first_begin_failed = true
+          raise ActiveRecord::ConnectionFailed, "Simulated failure"
+        end
+        super(raw_connection, sql, *args, **kwargs)
+      end
+
+      Sample.transaction(isolation: :read_committed) do
+        @connection.materialize_transactions
+        isolation = @connection.select_value("SELECT ISOLATION_LEVEL FROM performance_schema.events_transactions_current where THREAD_ID=PS_CURRENT_THREAD_ID()")
```
Comment (kirs) [inline activerecord/test/cases/adapters/abstract_mysql_adapter/transaction_test.rb:169]: This is MySQL 8 specific and looks to be be failing on CI :(

SELECT command denied to user 'rails'@'172.16.0.5' for table 'events_transactions_current'

Normally, you'd use something as simple as `select @@transaction_isolation` to check current isolation, but that one only works for session-wise transaction isolation setting.

**Comment by rafaelfranca** [comment]
The reason why we disallow multi-statements was to decrease the impact of SQL injections. I'm ok with us using for internal stuff like inserting fixtures or opening transaction, but I don't think we should allow to execute arbitrary queries using multi-statements

**Comment by byroot** [comment]
> I don't think we should allow to execute arbitrary queries using multi-statements

That's not what I'm suggesting.

**Comment by kirs** [comment]
@byroot we've worked around this by permanently setting isolation level on `mysqld`, but it would still be great to have this upstream 💚

**Comment by casperisfine** [comment]
Yeah, I'll poke Matthew, see if he has any strong opinions against this, if not I'll merge.

**Comment by casperisfine** [comment]
Discussed this with Matthew. We both agree that this is hinting at something more, like being able to have groups of queries retryable together etc, but since I've just went through a bit Active Record yak shave I'm not really willing to explore this just now.

Considering this, this solution is deemed acceptable for now. I'll fix CI and merge.

**Pull request merged**